service/src/main/java/com/commercetools/pspadapter/payone/notification/NotificationProcessorBase.java

```diff
@@ -53,29 +65,81 @@ public final void processTransactionStatusNotification(final Notification notifi
         }
 
         try {
-            client.executeBlocking(PaymentUpdateCommand.of(payment, createPaymentUpdates(payment, notification)));
+            // 1. update payment
+            CompletionStage<Order> fullCompletionStage = getPaymentService().updatePayment(payment, createPaymentUpdates(payment, notification))
+                    // 2. when payment is updated - try to find and updated respective Order.paymentState
+                    .thenComposeAsync(this::tryToUpdateOrderByPayment);
+
+            executeBlocking(fullCompletionStage);
+
         } catch (final io.sphere.sdk.client.ConcurrentModificationException e) {
             throw new java.util.ConcurrentModificationException(e);
         }
     }
 
     /**
+     * Fetch an order with respective payment id and try to update that order.
+     *
+     * @param updatedPayment <b>non-null</b> new updated payment instance.
+     * @return completion stage with nullable updated {@link Order}
+     */
+    private CompletionStage<Order> tryToUpdateOrderByPayment(Payment updatedPayment) {
+        return getOrderService().getOrderByPaymentId(updatedPayment.getId())
+                .thenComposeAsync(order -> updateOrderIfExists(order.orElse(null), updatedPayment));
+    }
+
+    /**
+     * If the {@code order} is not null and it's payment status is different from the new one - update it,
+     * otherwise - skip updating.
+     * <p>
+     * If a new payment status can't be mapped to the order's payment state - the value is left unchanged.
+     *
+     * @param order          <b>nullable</b> order to update state. If <b>null</b> - the update is skipped,
+     *                       but the incident is reported.
+     * @param updatedPayment <b>non-null</b> instance of the updated payment
+     * @return a {@link CompletionStage} with <b>nullable</b> updated {@link Order} instance. If incoming order is null
+     * or new payment state is undefined, or update is not required - the returned stage is completed and contains
+     * the same instance.
+     */
+    private CompletionStage<Order> updateOrderIfExists(Order order, Payment updatedPayment) {
+        if (order != null) {
```

Comment:
In this case i would not call ```updateOrderIfExists``` at all and call it rather ```updateOrder``` which is only called if order is found.

It seems that opinion of main developer is not shared at least in official article: http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html

> its purpose is to help design more-comprehensible APIs so that by just reading the signature of a method, you can tell whether you can expect an optional value. This forces you to actively unwrap an Optional to deal with the absence of a value

But lets not open holly-war. I am fine with null check also.